src/rest/activity.rs

```diff
@@ -0,0 +1,356 @@
+        )
+        .boxed_local()
+    }
+}
+
+pub struct SgxActivity {
+    secure_api: SecureActivityRequestorApi,
+    api: ActivityRequestorApi,
+    activity_id: String,
+    drop_list: CancelableDropList,
+}
+
+impl Drop for SgxActivity {
+    fn drop(&mut self) {
+        if let Some(ref drop_list) = self.drop_list.take() {
+            let api = self.api.clone();
+            let id = self.activity_id.clone();
+            drop_list.async_drop(async move {
+                api.control()
+                    .destroy_activity(&id)
+                    .await
+                    .with_context(|| format!("Failed to auto destroy Activity: {:?}", id))?;
+                log::debug!(target:"yarapi::drop", "Activity {:?} destroyed", id);
+                Ok(())
+            })
+        }
+    }
+}
+
+impl SgxActivity {
+    pub(crate) async fn create(
+        api: ActivityRequestorApi,
+        agreement_id: &str,
+        drop_list: CancelableDropList,
+    ) -> Result<Self> {
+        let secure_api = api
+            .control()
+            .create_secure_activity(agreement_id)
+            .await
+            .with_context(|| {
+                format!("failed to create activity for agreement {:?}", agreement_id)
+            })?;
+        let activity_id = secure_api.activity_id();
+
+        Ok(Self {
+            api,
+            secure_api,
+            activity_id,
+            drop_list,
+        })
+    }
+}
+
+impl Activity for SgxActivity {
+    type RunningBatch = SgxBatch;
+
+    fn id(&self) -> &str {
+        &self.activity_id
+    }
+
+    fn exec(
+        &self,
+        commands: Vec<ExeScriptCommand>,
+    ) -> LocalBoxFuture<'static, Result<Self::RunningBatch>> {
+        let api = self.secure_api.clone();
+        async move {
+            let batch_commands = commands.clone().into();
+            let batch_id = api.exec(commands).await?;
+            Ok(SgxBatch {
+                api,
+                batch_id,
+                commands: batch_commands,
+            })
+        }
+        .boxed_local()
+    }
+
+    fn credentials(&self) -> Option<Credentials> {
+        Some(self.secure_api.proof())
+    }
+
+    fn destroy(&self) -> LocalBoxFuture<'static, Result<()>> {
+        let api = self.api.clone();
+        let activity_id = self.activity_id.clone();
+        async move {
+            api.control()
+                .destroy_activity(&activity_id)
+                .await
+                .with_context(|| format!("failed to destroy sgx activity: {:?}", activity_id))
+        }
+        .boxed_local()
+    }
+}
+
+pub struct SgxBatch {
+    api: SecureActivityRequestorApi,
+    batch_id: String,
+    commands: Arc<[ExeScriptCommand]>,
+}
+
+impl RunningBatch for SgxBatch {
+    fn id(&self) -> &str {
+        &self.batch_id
+    }
+
+    fn events(&self) -> LocalBoxStream<'static, Result<Event>> {
+        let api = self.api.clone();
+        let batch_id: Arc<str> = self.batch_id.clone().into();
```

Comment:
Wouldn't just a `String` be a better choice here than `Arc<str>`? Unless cloning it will become insanely expensive? Do you have some idea how many clones we'd expect to warrant the use of `Arc<str>` instead of plain old `String`?

Reply:
Everything is a better choice then string. if you compare `rust` to other languages rust has the worst type for owned strings.
`&str` is good. but `String` ? in most languages you can pass a string in a single register. But rust string is `vec` with a triplet of (size, pointer, capacity). Arc<str> or Box<str> are 1/3 shorter. For example in java, you have `String` and `StringBuilder`. The first one is immutable and has fast substring. in rust as default every you have java's `StringBuffer`, why? maybe in most places `Box<str>` is the better choice than `String`.